### ISSUE TITLE
Non-orthogonal grid and illegal cells function

### ISSUE THREAD

**Issue by n-aleksandrova** [Deltares/dfm_tools#909]
Hi,

I am generating a grid using the latest release of dfm_tools (0.24.0), and using the dfmt.meshkernel_get_illegalcells method to identify non-orthogonal cells. However, I encountered cases where this function was not able to identify illegal cells, which is the case for a specific cell type, see snapshot below (where the blue circle shows the problematic cell that is not picked up by the identification method, and the red rectangle shows another type of problematic cell that is marked automatically as an illegal cell).
I was able to run the model after manually adding a polygon around the cells that caused the issue (my grid had 2 such cells), but would be great if this can be done automatically.

![image](https://github.com/user-attachments/assets/d6ff50b3-f45f-4d27-9efb-7e7887479953)

**Comment by veenstrajelmer** [comment]
Thanks for creating this issue. Would it be possible to share a minimal example that generates this part of the grid? It does make sense that this cell is not found, since it searches for cells with 6 edges, yours has only 4. However, I understand that this cell shape causes an issue, it should have an extra edge dividing it in two cells just like the similar cell a bit above right of this one.

**Comment by n-aleksandrova** [comment]
Indeed the missing edge is the problem, and I think this happens because of the rather complex coastline in that area. I tried to use the coastline with lower resolution, but that does not change the situation. In this particular grid I found two such cells with missing edges.

Below is the code used to generate the grid, let me know if you are missing any information.
```
import dfm_tools as dfmt
import xarray as xr

lon_min, lon_max, lat_min, lat_max =  32.55, 47.22, -27.84, -9.6
dxy = 0.1
min_edge_size = 400 # in meters

# generate spherical regular grid
mk_object = dfmt.make_basegrid(lon_min, lon_max, lat_min, lat_max, dx=dxy, dy=dxy, crs=crs)

# generate plifile from grid extent and coastlines
bnd_gdf = dfmt.generate_bndpli_cutland(mk=mk_object, res='h', buffer=0.2)
bnd_gdf_interp = dfmt.interpolate_bndpli(bnd_gdf, res=0.03)

# filter out boundary sections that are very short (<5km in this case)
ids = []
for ii in range(len(bnd_gdf_interp.index)):
    if bnd_gdf_interp.iloc[ii].values.to_crs('EPSG:3857').length < 5000: # approx. 111 km in 1 degree
        ids.append(ii)

bnd_gdf_interp = bnd_gdf_interp.drop(ids)

file_nc_bathy_sel = r'p:\11210471-001-compass\02_Models\Delft3DFM\mozambique_model\boundary_conditions\GEBCO_MZB\gebco_2023_n-8.0_s-29.0_w31.5_e48.5.nc'
data_bathy_sel = xr.open_dataset(file_nc_bathy_sel).elevation
data_bathy_sel.load()

# refine grid
dfmt.refine_basegrid(mk=mk_object, data_bathy_sel=data_bathy_sel, min_edge_size=min_edge_size)

# remove land with GSHHS coastlines
dfmt.meshkernel_delete_withcoastlines(mk=mk_object,res='i',min_area=50,crs=crs)

# derive illegalcells geodataframe
illegalcells_gdf = dfmt.meshkernel_get_illegalcells(mk=mk_object)

# convert to xugrid
xu_grid_uds = dfmt.meshkernel_to_UgridDataset(mk=mk_object, crs=crs)

# interpolate bathymetry onto the grid
data_bathy_interp = data_bathy_sel.interp(lon=xu_grid_uds.obj.mesh2d_node_x, lat=xu_grid_uds.obj.mesh2d_node_y)
xu_grid_uds['mesh2d_node_z'] = data_bathy_interp.clip(max=10)

# write xugrid grid to netcdf
netfile = os.path.join(dir_output_run, f'{model_name}_net.nc')
xu_grid_uds.ugrid.to_netcdf(netfile)
```

**Comment by veenstrajelmer** [comment]
Thanks, I have minimized your code so the issue can be easily reproduced:
```python

import dfm_tools as dfmt
import xarray as xr
import matplotlib.pyplot as plt
plt.close("all")

# lon_min, lon_max, lat_min, lat_max =  32.55, 47.22, -27.84, -9.6
lon_min, lon_max, lat_min, lat_max =  46.2, 46.8, -18, -15.85
dxy = 0.1
min_edge_size = 400 # in meters
crs = "EPSG:4326"

# generate spherical regular grid
mk_object = dfmt.make_basegrid(lon_min, lon_max, lat_min, lat_max, dx=dxy, dy=dxy, crs=crs)

# generate plifile from grid extent and coastlines
bnd_gdf = dfmt.generate_bndpli_cutland(mk=mk_object, res='h', buffer=0.2)
bnd_gdf_interp = dfmt.interpolate_bndpli(bnd_gdf, res=0.03)

file_nc_bathy_sel = r'p:\11210471-001-compass\02_Models\Delft3DFM\mozambique_model\boundary_conditions\GEBCO_MZB\gebco_2023_n-8.0_s-29.0_w31.5_e48.5.nc'
data_bathy_sel = xr.open_dataset(file_nc_bathy_sel).elevation
data_bathy_sel = data_bathy_sel.sel(lon=slice(lon_min-0.1,lon_max+0.1),
                                    lat=slice(lat_min-0.1,lat_max+0.1))
data_bathy_sel.load()

# refine grid
dfmt.refine_basegrid(mk=mk_object, data_bathy_sel=data_bathy_sel, min_edge_size=min_edge_size)

# remove land with GSHHS coastlines
# dfmt.meshkernel_delete_withcoastlines(mk=mk_object,res='i',min_area=50,crs=crs)
mesh_bnds = mk_object.mesh2d_get_mesh_boundaries_as_polygons()
bbox = (mesh_bnds.x_coordinates.min(), mesh_bnds.y_coordinates.min(), mesh_bnds.x_coordinates.max(), mesh_bnds.y_coordinates.max())
coastlines_gdf = dfmt.get_coastlines_gdb(bbox=bbox, res='i', min_area=50, crs=crs)
dfmt.meshkernel_delete_withgdf(mk=mk_object, coastlines_gdf=coastlines_gdf)

# derive illegalcells geodataframe
illegalcells_gdf = dfmt.meshkernel_get_illegalcells(mk=mk_object)

# convert to xugrid
xu_grid_uds = dfmt.meshkernel_to_UgridDataset(mk=mk_object, crs=crs)

# add orthogonality
ortho = mk_object.mesh2d_get_orthogonality()
ortho_vals = ortho.values
ortho_vals[ortho_vals==ortho.geometry_separator] = 0
print("ortho max:", ortho_vals.max())
xu_grid_uds['ortho'] = xr.DataArray(ortho_vals, dims=xu_grid_uds.grid.edge_dimension)

# plot TODO: the order of edges/orthogonality is not aligned yet in xugrid 0.10.0, but only after https://github.com/Deltares/xugrid/issues/148 is fixed (and released)
fig, ax = plt.subplots()
xu_grid_uds['ortho'].ugrid.plot(ax=ax)
coastlines_gdf.plot(ax=ax, color="none", edgecolor='r')
```

Gives:
![image](https://github.com/user-attachments/assets/0bb9dc39-0293-43b6-95e3-e2ff37d21da7)

The issue is that the two cells are marked as to be deleted by the meshkernel deletion algorithm. This is an edge case that I hope we can avoid in the future. For now, you can manually add illegalcells, but you could also try using `res='h'` in `dfmt.meshkernel_delete_withcoastlines()`. It solves this specific cell, but it might still result in other similar cells.

**Comment by lucacarniato** [comment]
In the upcoming release, a new API will be introduced to retrieve face polygons with all edges having orthogonality values falling within a specified range. For an example, please take a look at the image below.

![image](https://github.com/user-attachments/assets/d83023fe-e7cb-4be5-b0c7-a683d6d72bd3)

**Comment by veenstrajelmer** [comment]
Implemented in https://github.com/Deltares/MeshKernel/pull/367, also call this function in dfm_tools

**Comment by famkekovacs-deltares** [comment]
Hi, a new issue has arrived that the illegalcells_gdf is empty, however the orthogonaliy of some points is still near 1 if you plot them. Here is the code I used
`# -*- coding: utf-8 -*-
"""
Created on Wed Jan 22 12:25:34 2025

@author: kovacs
"""

import dfm_tools as dfmt
import xarray as xr
import matplotlib.pyplot as plt
plt.close("all")

lon_min, lon_max, lat_min, lat_max =  -1, 2.5, 52.5, 55

dxy = 0.05
min_edge_size = 300 # in meters
crs = "EPSG:4326"

# generate spherical regular grid
mk_object = dfmt.make_basegrid(lon_min, lon_max, lat_min, lat_max, dx=dxy, dy=dxy, crs=crs)

# generate plifile from grid extent and coastlines
bnd_gdf = dfmt.generate_bndpli_cutland(mk=mk_object, res='h', buffer=0.01)
bnd_gdf_interp = dfmt.interpolate_bndpli(bnd_gdf, res=0.05)

file_nc_bathy_sel = 'https://opendap.deltares.nl/thredds/dodsC/opendap/deltares/Delft3D/netcdf_example_files/GEBCO_2022/GEBCO_2022_coarsefac08.nc'
data_bathy_sel = xr.open_dataset(file_nc_bathy_sel).elevation
data_bathy_sel = data_bathy_sel.sel(lon=slice(lon_min-0.1,lon_max+0.1),
                                    lat=slice(lat_min-0.1,lat_max+0.1))
data_bathy_sel.load()

# refine grid
dfmt.refine_basegrid(mk=mk_object, data_bathy_sel=data_bathy_sel, min_edge_size=min_edge_size)

# remove land with GSHHS coastlines
dfmt.meshkernel_delete_withcoastlines(mk=mk_object, res='h')

# derive illegalcells geodataframe
illegalcells_gdf = dfmt.meshkernel_get_illegalcells(mk=mk_object)

# convert to xugrid
xu_grid_uds = dfmt.meshkernel_to_UgridDataset(mk=mk_object, crs=crs)

# add orthogonality
ortho = mk_object.mesh2d_get_orthogonality()
ortho_vals = ortho.values
ortho_vals[ortho_vals==ortho.geometry_separator] = 0
print("ortho max:", ortho_vals.max())
print('mk.mesh2d_get_orthogonality()')
print(mk_object.mesh2d_get_orthogonality().values.max())
xu_grid_uds['ortho'] = xr.DataArray(ortho_vals, dims=xu_grid_uds.grid.edge_dimension)

# plot NOTE: the order of edges/orthogonality is correctly aligned from xugrid 0.11.0
fig, ax = plt.subplots()
xu_grid_uds['ortho'].ugrid.plot(ax=ax)
mk_object.mesh2d_get().plot_edges(ax,zorder=1)`

**Comment by veenstrajelmer** [comment]
According to @arthurvd the face-node-connectivity is read by dflowfm if it is supplied in the network file. So it is not recomputed in that case: https://git.deltares.nl/oss/delft3d/-/blob/main/src/engines_gpl/dflowfm/packages/dflowfm_kernel/src/dflowfm_kernel/prepost/flow_geominit.f90#L237-245
That would mean these type of holes should not cause orthogonality errors in dflowfm.

**Comment by veenstrajelmer** [comment]
Example that generates illegalcells without dfm_tools dependency:
```python
import matplotlib.pyplot as plt
plt.close('all')
import numpy as np
from meshkernel import (ProjectionType, MakeGridParameters, MeshKernel, 
                        MeshRefinementParameters, GriddedSamples, RefinementType,
                        GeometryList, DeleteMeshOption,
                        )

# input params
lon_min, lon_max, lat_min, lat_max = 147.75, 147.9, -40.4, -40.25
dxy = 0.05 #0.05
crs = 'EPSG:4326'

# grid generation and refinement with GEBCO bathymetry
# create base grid
projection = ProjectionType.SPHERICAL
make_grid_parameters = MakeGridParameters(angle=0,
                                            origin_x=lon_min,
                                            origin_y=lat_min,
                                            upper_right_x=lon_max,
                                            upper_right_y=lat_max,
                                            block_size_x=dxy,
                                            block_size_y=dxy)

mk = MeshKernel(projection=projection)
mk.curvilinear_compute_rectangular_grid_on_extension(make_grid_parameters)
mk.curvilinear_convert_to_mesh2d() #convert to ugrid/mesh2d

#refine
min_edge_size = 500 #in meters
print('Refining basegrid')
gebco_elev = np.array([[-37, -37, -36, -36],
        [-41, -39, -38, -34],
        [-42, -32, -27,  -6],
        [-38,  -6,  -2,   2],
        [-43, -42, -31, -20]],dtype=np.int16)
lat_np = np.array([-40.397917, -40.364583, -40.33125 , -40.297917, -40.264583])
lon_np = np.array([147.76875 , 147.802083, 147.835417, 147.86875 ])
values_np = gebco_elev.flatten()
gridded_samples = GriddedSamples(x_coordinates=lon_np,y_coordinates=lat_np,values=values_np)


#refinement
mesh_refinement_parameters = MeshRefinementParameters(min_edge_size=min_edge_size, #always in meters
                                                                 refinement_type=RefinementType(1), #Wavecourant/1,
                                                                 connect_hanging_nodes=True, #set to False to do multiple refinement steps (e.g. for multiple regions)
                                                                 smoothing_iterations=2,
                                                                 max_courant_time=120)

mk.mesh2d_refine_based_on_gridded_samples(
    gridded_samples=gridded_samples,
    mesh_refinement_params=mesh_refinement_parameters,
    )

print("max ortho after refining", mk.mesh2d_get_orthogonality().values.max())

#cutcells
geometry_separator = -999
xx1, yy1 = np.array([
    [147.84157258, -40.30590657],
       [147.84157258, -40.29819399],
       [147.85931452, -40.29728663],
       [147.85754032, -40.30590657],
       [147.84157258, -40.30590657],
       ]).T
xx2, yy2 = np.array([
    [147.87262097, -40.33993264],
       [147.89391129, -40.33993264],
       [147.90366935, -40.36533877],
       [147.87084677, -40.35490411],
       [147.87262097, -40.33993264],
       ]).T
xx = np.concatenate([xx1, [geometry_separator], xx2])
yy = np.concatenate([yy1, [geometry_separator], yy2])
delete_pol_geom = GeometryList(x_coordinates=xx, y_coordinates=yy, geometry_separator=geometry_separator) #TODO: .copy()/to_numpy() makes the array contiguous in memory, which is necessary for meshkernel.mesh2d_delete()
mk.mesh2d_delete(geometry_list=delete_pol_geom, 
                 delete_option=DeleteMeshOption.INSIDE_NOT_INTERSECTED,
                 invert_deletion=False)

ortho = mk.mesh2d_get_orthogonality().values
print("max ortho after cutting",ortho.max())

m2d = mk.mesh2d_get()
bool_ortho_high = ortho>0.4
sel_x = m2d.edge_x[bool_ortho_high]
sel_y = m2d.edge_y[bool_ortho_high]
sel_o = ortho[bool_ortho_high]

fig, ax = plt.subplots()
m2d.plot_edges(ax=ax,linewidth=1)
ax.scatter(sel_x,sel_y,5,c=sel_o, zorder=5)
ax.plot(xx1, yy1, "r-")
ax.plot(xx2, yy2, "r-")
```

Gives:

<img width="577" height="412" alt="Image" src="https://github.com/user-attachments/assets/5b2b7fbf-d340-48f1-882e-cd1adc85b24a" />

**Comment by veenstrajelmer** [comment]
With wheel built on 27 Oct 17.08, the findings are different. The meshkernel-code now shows that illegalcells are being correctly interpreted as holes in the grid (green bounds and excluded from orthogonality computation):
<img width="577" height="410" alt="Image" src="https://github.com/user-attachments/assets/101f5dfa-9daa-4cba-8a7f-8943fea3c887" />

The original dfm_tools code also shows the same improvement (and runs in 10-12 seconds):
<img width="734" height="406" alt="Image" src="https://github.com/user-attachments/assets/cb37eb1f-849b-40ff-a405-9b4628725ad9" />

The latter code runs in 10-12 seconds. The original code runs in approximately the same time, but it is wise to do some additional performance checks.
